### PR TITLE
🚧👌 Link repo to cookiecutter template with cruft

### PR DESCRIPTION
**Pull request by s-weigand**
This PR links the project with the [upstream cookiecutter template](https://github.com/s-weigand/cookiecutter-pypackage) it was created with, allowing for easier propagation of changes and improved maintainability.

## Notable changes:
- ✨ Add official python 3.12 support
- 💥 Drop python 3.8 support (will be EOL 31 Oct 2024 and was already dropped by `PySide6==6.7.2`)
- 🧰👌 Use ruff for lining and formatting
- ⌨️👌 Use modern python 3.10 typing style
- 🧰 Use `pyptoject.toml` for package definition and hatch for management
- 👌 Use importlib-resources files API instead of deprecated `importlib.resources.path`
- 🚇🩹 Add codecov token so codecov works again

## Summary by Sourcery

Link the project to the upstream cookiecutter template using cruft, add support for Python 3.12, and drop support for Python 3.8. Transition to using `pyproject.toml` and hatch for project management, and adopt modern Python typing styles. Replace deprecated APIs and update linting and formatting tools to use Ruff. Enhance CI workflows and documentation, and refactor tests for improved maintainability.

New Features:
- Add support for Python 3.12 and drop support for Python 3.8, aligning with the end-of-life schedule and compatibility with PySide6.
- Integrate the project with the upstream cookiecutter template using cruft for easier updates and maintainability.

Enhancements:
- Switch to using `pyproject.toml` for package definition and hatch for project management.
- Adopt modern Python 3.10 typing style across the codebase.
- Replace deprecated `importlib.resources.path` with `importlib-resources` files API.
- Use Ruff for linting and formatting, replacing previous tools like Black and Flake8.

Build:
- Update pre-commit configuration to include new hooks and remove outdated ones.
- Modify dependencies installation in CI workflows to use `requirements_pinned.txt` and `.[docs]`.

CI:
- Add a Codecov token to ensure coverage reports are uploaded correctly.
- Update GitHub Actions workflows to support Python 3.12 and remove Python 3.8.

Documentation:
- Add new documentation files including installation, usage, and contributing guidelines.
- Update README to reflect changes in code style tools and add a new FAQ section.

Tests:
- Refactor tests to use modern Python features and improve parameterization for better readability and maintainability.

Chores:
- Add a `.cruft.json` file to track the link with the cookiecutter template.
- Remove `setup.cfg` in favor of `pyproject.toml` for configuration.

### CODE REVIEW

File: tests/cli/test_main_app.py [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
-from _pytest.monkeypatch import MonkeyPatch
 from typer.testing import CliRunner
 
 from qt_dev_helper.cli.main_app import app
 
 
-def test_missing_cli_extra_requires(monkeypatch: MonkeyPatch):
-    """Exception raised if cli extra_requires is missing"""
+def test_missing_cli_extra_requires(monkeypatch: pytest.MonkeyPatch):
+    """Exception raised if cli extra_requires is missing."""
```
[the remark this answers is not in the document]
Comment (s-weigand) [reply tests/cli/test_main_app.py:15]: The positive test is the default.

File: tests/cli/test_utils.py [hunk 2]
```diff
@@ -8,14 +8,14 @@
 
 
 @pytest.mark.parametrize(
-    "optional_args_string, expected",
-    (
+    ("optional_args_string", "expected"),
+    [
```
[the remark this answers is not in the document]
Comment (s-weigand) [reply tests/cli/test_utils.py:12]: The correct type is ensured by typer.

File: tests/conftest.py [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
 
 from qt_dev_helper.config import Config
+from tests import TEST_DATA
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+    from pathlib import Path
 
 
 @pytest.fixture
```
[the remark this answers is not in the document]
Comment (s-weigand) [reply tests/conftest.py:19]: The fixuture value is updated as needed in tests.

File: tests/test_utils.py [hunk 4]
```diff
@@ -31,11 +31,11 @@ def test_format_rel_output_path(format_string: str, flatten_path: bool, expected
...
-    (
+    ("recurse_folder", "max_index"),
+    [
         (True, 4),
         (False, 2),
-    ),
+    ],
 )
 def test_find_matching_files(
     nested_ui_folder: tuple[Path, Path, Path, Path], recurse_folder: bool, max_index: int
```
[the remark this answers is not in the document]
Comment (s-weigand) [reply tests/test_utils.py:41]: The function uses glob, which is a noop generator for an empty directory so no need for this kind of test.

**Review by jsnel — APPROVED** [review]
Reviewed ok!

**Pull request merged**